src/backend/lib/upgrade-manager.js

```diff
@@ -1,192 +1,275 @@
+    /** @private */
+    this._storage = new UpgradeStorage({ storageDir, currentApkInfo });
+    /** @private */
+    this._server = new UpgradeServer({ storage: this._storage });
+    /** @private */
+    this._discovery = new UpgradeDiscovery({
+      discoveryKey: currentApkInfo.applicationId,
     });
 
-    // State aggregation
-    this.state = {
-      server: { state: this.server.state, context: this.server.context },
-      downloader: {
-        search: {
-          state: this.downloader._search.state,
-          context: this.downloader._search.context,
-        },
-        download: {
-          state: this.downloader._download.state,
-          context: this.downloader._download.context,
-        },
-        check: {
-          state: this.downloader._check.state,
-          context: this.downloader._check.context,
-        },
-      },
-    };
-    this.server.on("state", (state, context) => {
-      this.state.server = { state, context };
-      this.emitState();
+    this._server.on("uploads", uploads => {
+      this.setState({ uploads });
+    });
+    this._server.on("error", error => {
+      this.emit("error", error);
+    });
+    this._storage.on("error", error => {
+      this.setState({ value: "error", error });
+    });
+    this._storage.on("installers", async installers => {
+      this._onStoredInstallers(installers);
     });
-    this.downloader.on("state", state => {
-      this.state.downloader = state;
-      this.emitState();
+    this._discovery.on("installers", async installers => {
+      this._onDownloadableInstallers(installers);
     });
+    this.throttledEmitState = throttle(() => {
+      this.emit("state", this.getState());
+    }, EMIT_THROTTLE);
   }
 
-  // Ensures that any Error instances are converted into objects that serialize
-  // properly over the RN Bridge. Mutates state.
-  sanitizeState() {
-    if (this.state.server.context instanceof Error) {
-      this.state.server.context = {
-        message: this.state.server.context.message,
-      };
-    }
-    if (this.state.downloader.search.context instanceof Error) {
-      this.state.downloader.search.context = {
-        message: this.state.downloader.search.context.message,
-      };
-    }
-    if (this.state.downloader.download.context instanceof Error) {
-      this.state.downloader.download.context = {
-        message: this.state.downloader.download.context.message,
-      };
-    }
-    if (this.state.downloader.check.context instanceof Error) {
-      this.state.downloader.check.context = {
-        message: this.state.downloader.check.context.message,
-      };
-    }
+  /**
+   * Called whenever the list of installers in local storage changes. If there
+   * is an upgrade candidate in storage, then this will emit state with that
+   * upgrade candidate
+   *
+   * @private
+   * @param {InstallerInt[]} storedInstallers
+   */
+  _onStoredInstallers(storedInstallers) {
+    const upgradeCandidate = getUpgradeCandidate({
```

Comment:
I think that `getBestUpgradeCandidate` could be a more instructive name for this helper function?

Reply:
Agreed! Changed.